os/fs/vfs/fs_truncate.c

```diff
@@ -0,0 +1,211 @@
+	 * a write-able file system.
+	 */
+
+	if (inode->u.i_mops->truncate == NULL) {
+		fwdbg("WARNING: File system does not support the truncate() method\n");
+		return -ENOSYS;
+	}
+
+	/* Yes, then tell the file system to truncate this file */
+
+	return inode->u.i_mops->truncate(filep, length);
+}
+
+/****************************************************************************
+ * Name: ftruncate
+ *
+ * Description:
+ *   The ftruncate() function causes the regular file referenced by fd to
+ *   have a size of length bytes.
+ *
+ *   If the file previously was larger than length, the extra data is
+ *   discarded.  If it was previously shorter than length, it is unspecified
+ *   whether the file is changed or its size increased.  If the file is
+ *   extended, the extended area appears as if it were zero-filled.  If fd
+ *   references a shared memory object, ftruncate() sets the size of the
+ *   shared memory object to length. If the file is not a regular file or
+ *   a shared memory object, the result is unspecified.
+
+ *   With ftruncate(), the file must be open for writing; for truncate(),
+ *   the process must have write permission for the file.
+ *
+ *   ftruncate() does not modify the file offset for any open file
+ *   descriptions associated with the file.
+ *
+ * Input Parameters:
+ *   fd     - A reference to an open, regular file or shared memory object
+ *            to be truncated.
+ *   length - The new length of the file or shared memory object.
+ *
+ * Returned Value:
+ *    Upon successful completion, ftruncate() return 0s. Otherwise a -1 is
+ *    returned, and errno is set to indicate the error.
+ *
+ *    EINTR
+ *      - A signal was caught during execution.
+ *    EINVAL
+ *      - The length argument was less than 0.
+ *    EFBIG or EINVAL
+ *      - The length argument was greater than the maximum file size.
+ *    EIO
+ *      - An I/O error occurred while reading from or writing to a file
+ *        system.
+ *    EBADF or EINVAL
+ *       - the fd argument is not a file descriptor open for writing.
+ *    EFBIG
+ *       - The file is a regular file and length is greater than the offset
+ *         maximum established in the open file description associated with
+ *         fd.
+ *    EINVAL
+ *       - The fd argument references a file that was opened without write
+ *         permission.
+ *    EROFS
+ *       - The named file resides on a read-only file system.
+ *
+ ****************************************************************************/
+
+int ftruncate(int fd, off_t length)
+{
+	FAR struct file *filep;
+	int ret;
+
+	/* Get the file structure corresponding to the file descriptor. */
+
+	filep = fs_getfilep(fd);
+	if (filep == NULL) {
+		fdbg("ERROR: Could no get file structure\n");
+		goto errout;
+	}
+
+	DEBUGASSERT(filep != NULL);
+
+	/* Perform the truncate operation */
+
+	ret = file_truncate(filep, length);
+	if (ret >= 0) {
+		return 0;
```

Comment:
no, errno should be set for each of file operation.

Reply:
This line represents success case.